### ISSUE TITLE
Should we support native scrolling?

### ISSUE THREAD

**Issue by idiotWu** [idiotWu/smooth-scrollbar#27]
As is explained in issue #14, smooth-scrollbar breaks native scrolling experience on some devices (e.g. iOS/Android with hardware accelerated scrolling). So I am thinking about the possibility of integrating with native scrollbar.

The following is a table of API compatibility with native scrollbar:

| API name | Compatibility | Explanation |
| :------ | :------- | :------ |
| [instance#targets](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancetargets) | ✔️ (partial) | We may no longer need `scroll-content` wrapper and `scrollbar-track` |
| [instance#offset](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceoffset) | ✔️ | As an accessor property that returns `el.scrollTop` and `el.scrollLeft`. |
| [instance#limit](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancelimit) | ✔️ | As(maybe) an accessor property. | 
| [instance#scrollTop](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancescrolltop) | ✔️ | |
| [instance#scrollLeft](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancescrollleft) | ✔️ | |
| [instance#update()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceupdate-async-) | ❌ | Unnecessary. |
| [instance#unregisterEvents()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceunregisterevents-regex--regex-regex--) | ❌ | Unnecessary. |
| [instance#registerEvents()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceregisterevents-regex--regex-regex--) | ❌ | Unnecessary. |
| [instance#clearMovement()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#clearMovement) | ❌ | Unnecessary. |
| [instance#stop()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#stop) | ❌ | Unnecessary. |
| [instance#destroy()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancedestroy) | ❌ | Unnecessary. |
| [instance#showTrack()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceshowtrack-direction-) | ❌ | Unnecessary and impossible. |
| [instance#hideTrack()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancehidetrack-direction-) | ❌ | Unnecessary and impossible. |
| [instance#setPosition()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancesetposition-x-y-withoutcallbacks-) | ✔️(partial)  | `withoutCallbacks` parameter may be broken. |
| [instance#getSize()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancegetsize) | ✔️ | |
| [instance#scrollTo()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancescrollto-x-y-duration-callback-) | ✔️ | |
| [instance#scrollIntoView()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instancescrollintoview-elem-options-) | ✔️ | |
| [instance#isVisible()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceisvisible-elem-) | ✔️ | |
| [instance#addListener()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceaddlistener-fn-) | ✔️ | |
| [instance#removeListener()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceremovelistener-fn-) | ✔️ | |
| [instance#infiniteScroll()](https://github.com/idiotWu/smooth-scrollbar/wiki/Instance-Methods#instanceinfinitescroll-callback-threshold-) | ✔️ | |

**Comment by idiotWu** [comment]
Well...the compatibility looks nice. However it may be a tough work to make the integration, and file size will increase by at lease 30% I think.

In my view we'd better not support native momentum scrolling, it's complex and useless. Users who want to perform native scrolling, just throw this plugin away and listen to original `scroll` event. You can not alway cover all situations.

What's your option @bfred-it?

**Comment by bfred-it** [comment]
The problem with your last paragraph is who you mean by users. If you make it optional, the developers will be the users and most devs won't really care about degraded experience unless they themselves use these devices. The actual users of the scrolling (website visitors) will be the ones with a poor experience... by default.

`scrollbar-track` can and probably should stay, to keep the style consistent, so the track-related methods could also keep working.

**Comment by idiotWu** [comment]
@bfred-it so I think you may not be a potential user/developer of this script.

This plugin (and many others) is designed for those who want to **customize** scrollbars on all kinds of browsers, css is too weak to do this so we need a javascript simulation.

On desktop devices, scrolling is based on wheel delta from your input devices (mouse, trackpad etc.), if `friction` is set to `100%`, easing effect will be disabled and the behavior is absolutely the native one.

**Comment by bfred-it** [comment]
Again, it doesn't matter to me as a developer but as the user. As a developer I already know I can customize scrollbars with CSS [in the majority of all browsers](http://caniuse.com/#feat=css-scrollbar) and other JS solutions that don't break the native scroll _by default._

The problem is with the smoothing part. `damping:1` feels good, actually. Would it make sense to set it automatically once native smooth scrolling is (somehow) detected?

**Comment by idiotWu** [comment]
@bfred-it I think it's impossible to detect whether it's smooth scrolling. The only thing I can do is to use `damping < 1` to apply our smooth scrolling on most devices, or `damping = 1` to switch to native scrolling on particular devices (e.g. OSX with trackpad), it's a trade-off.

I am not able to tell which kind of input device it is, and I have no idea to set `damping` value automatically according to current environment. So I made it a configurable option and leave it up to developers. 

Users in your words suffer from poor scrolling because I (or developers) want them to behave the same way unless we could distinguish between the input hardwares, it's rude, it's communist, but it's a trade-off.

**Comment by bfred-it** [comment]
> _A communist lives in that house!_

You're free to be a communist but know that you're contributing to making the web worse when you don't have to.

http://codepen.io/bfred-it/full/pboNvw/

This is my graph for a big flick and a medium flick:

<img width="415" alt="screen shot 2016-05-28 at 15 04 44" src="https://cloud.githubusercontent.com/assets/1402241/15626089/2fc764dc-2470-11e6-97d7-1282755e93f7.png">

I'm not entirely sure of how it appears on non-smooth devices. While you might not be able to tell whether the device has smooth scrolling in the first scroll event, you can easily tell after they let go and the momentum keeps it going down smoothly.

**Comment by idiotWu** [comment]
>  you can easily tell after they let go...

@bfred-it I think it's hard to detect the release action. Maybe I can record an amount of momentum and compare it with the mathematical model of smooth scrolling - wow am I building an A.I.?

BTW my movement graph is similar to yours:

<img width="319" alt="2016-05-28 5 27 52" src="https://cloud.githubusercontent.com/assets/6022672/15626206/900bff4a-24f9-11e6-916d-3f6934670439.png">

**Comment by bfred-it** [comment]
No, use the codepen example, I'm not trying to show how comparable they are but if we can detect the native smooth scrolling

**Comment by idiotWu** [comment]
@bfred-it yeah I know what you're arguing about, but:

>  it's hard to detect the release action

So I can't:

> easily tell after they let go and the momentum keeps it going down smoothly

**Comment by bfred-it** [comment]
The plot looks substantially different; in your stepped plot there's no slope, the delta switches between nearly identical large steps, rather than decreasingly small steps. I think it would be _as easy as:_
```js
d = lastFiveDeltas;
d[0] - d[1] < 
d[1] - d[2] < 
d[2] - d[3] < 
d[3] - d[4] 
```

In short, if the delta's delta is less than the previous couple, then it's reasonably a momentum scroll.

This check could run at the beginning, maybe for the first seconds or until it matches.

**Comment by idiotWu** [comment]
@bfred-it a little attempt on your theory: http://codepen.io/idiotWu/full/YWzpQz/

As you can see, the result is unstable.

**Comment by bfred-it** [comment]
It doesn't need to be stable, once it detects it once or twice it should stop and set `damping:1`: 

<img width="386" alt="screen shot 2016-05-28 at 16 50 10" src="https://cloud.githubusercontent.com/assets/1402241/15626648/ed500a8c-247e-11e6-9e0f-ae4f2aba0f96.png">

How does it look with a stepped scrollwheel?

**Comment by bfred-it** [comment]
Either way I'd at least start disabling it almost entirely on mobile since:

* the scroll is always smooth
* mobile devices essentially don't have a scrollbar
* performance is guaranteed to be worsened and lead to crashes when you apply `translate3d` to long pages, especially when the devices is not the latest.

**Comment by idiotWu** [comment]
@bfred-it the logs of stepped mousewheel:

```
false, 288,300,308,304,300, delta's delta: 12,8,-4,-4
false, 300,308,304,300,240, delta's delta: 8,-4,-4,-60
false, 240,200,156,92,56, delta's delta: -40,-44,-64,-36
false, 200,156,92,56,20, delta's delta: -44,-64,-36,-36
true, 156,92,56,20,4, delta's delta: -64,-36,-36,-16
```

Look at `true` on line 5.

**Comment by bfred-it** [comment]
Right, then look at the delta's delta, it's one more check: `delta < 3`

**Comment by idiotWu** [comment]
@bfred-it yes I agree with you on the point of mobile devices, that's the reason why I opened this issue :)
